invisible_cities/reco/krmap_functions.py

```diff
@@ -149,3 +157,326 @@ def transform_parameters(fit_output : FitFunction):
+    -------
+    n_bins: int
+        Number of bins in each direction (X,Y) (square map).
+    '''
+    if    n_bins is not None: return n_bins;
+    elif  nevents < thr: return np.array([50, 50]);
+    else: return  np.array([100, 100]);
+
+
+def get_bin_counts_and_event_bin_id(dst    : pd.DataFrame,
+                                    bins_x : np.array,
+                                    bins_y : np.array):
+    '''
+    This function distributes all the events in the DST into the selected
+    bins. Given a certain binning, it computes which events fall into each
+    square of the grid formed by the bins arrays. binned_statistic_2d returns
+    counts (matrix here each matrix element is the number of events falling
+    into that specific bin) and bin_indexes (a 2-D array labeling each event
+    with the bin it belongs). Then counts is flattened into 1-D, bin_indexes
+    is transformed into 1-D using the number of bins on each axis.
+
+    Parameters
+    ----------
+    dst    : pd.DataFrame
+         Krypton dataframe.
+    bins_x : np.array
+         Binning in x axis
+    bins_y : np.array
+         Binning in y axis
+
+    Returns
+    -------
+    counts     : np.array
+         Total number of events falling into each bin
+    bin_labels : np.array
+         1D bin label for each event
+
+    Further info:
+    -------------
+    Why set expand_binnumbers to True (2D binning) if then we transform it to 1D?
+    Because even though expand_binnumbers = False returns 1-D labels, it also adds
+    two additional bins (per axis), taking into account out-of-range events which
+    dont fall into the the binning passed to the binned_statistic_2d function. But
+    since the dst is going to be previously selected and filtered with the desired
+    binning, it's not convenient to use that. Maybe a visual example is more useful:
+
+    2x2 binning (4 bins), natural index values shown both as 1D and 2D:
+
+    || 0 | 1 ||          || (0, 0) | (0, 1) ||
+    || 2 | 3 || (1D)  =  || (1, 0) | (1, 1) || (2D)
+
+    Using expand_binnumbers = False, the 1D index values instead of (0, ..., 3)
+    would be (0, ..., 15):
+
+    || 0 | 1 | 2 | 3 ||
+    || 4 | 5 | 6 | 7 ||  The bins that we "care about" (inner part) have indexes
+    || 8 | 9 |10 |11 ||  5, 6, 9, 10 which I believe is not convenient at all.
+    ||12 |13 |14 |15 ||  This creates (nx+2)*(ny+2) bins.
+    '''
+    n_xbins    = len(bins_x)-1
+    n_ybins    = len(bins_y)-1
+
+    counts, _, _, bin_indexes = stats.binned_statistic_2d(x=dst.X, y=dst.Y, values=None,
+                                                          bins=[bins_x, bins_y], statistic='count',
+                                                          expand_binnumbers=True)
+
+    counts       = counts.flatten()
+    bin_indexes -= 1
+    bin_labels   = np.ravel_multi_index(bin_indexes, dims=(n_xbins, n_ybins),
+                                        mode='clip', order = 'F')
+
+    return counts, bin_labels
```

Comment:
```suggestion
    # indexes 0 and len+1 represent underflow and overflow, thus we subtract 1
    bin_labels   = np.ravel_multi_index(bin_indexes - 1, dims=(n_xbins, n_ybins),
                                        mode='clip', order = 'F')

    return counts.flatten(), bin_labels
```

Reply:
if I understand correctly `mode="clip"` means that bins out of range will be assigned the closest bin. Is this what we used to do @bpalmeiro ?

Reply:
In our case, we had three values for "fiducial":

- The R max for the selection (just a quality one to ensure everything was inside)

- The R max for the map, that defines the maximum extension of it. In this case, if the event didn't happen to fall inside the bins, it was disregarded.

- Also (not sure here it's the most appropriate place to mention, tho), we had a posterior step where the peripheral bins (those further than a fiducial r -but within said Rmax for map production-) were set to nan.

The latter was used to ensure that, in the posterior analysis, all the hits reconstructed outside the volume were automatically flagged as nan.

Reply:
That said, the best approach would be to disregard events outside boundaries instead of fakely merging them in the border bins, even if we rely on a previous selection. It's not a very strong opinion, but I think it adds redundancy (and, therefore, a bit of robustness) to the process.

Reply:
Yes, my bad... I assumed that there would not be any events outside the boundaries here since I was expecting a "clean" `dst` in this part of the code, so when I defined `bin_labels` like that I didn't think of the consequences in case some event was actually out of the range.

I am thinking in a solution like this one: 

```suggestion
    counts       = counts.flatten()
    bin_indexes -= 1
    valid_mask = ( in_range(bin_indexes[0], 0, n_xbins, right_closed=True) &
                   in_range(bin_indexes[1], 0, n_ybins, right_closed=True) )

    bin_labels             = np.full(bin_indexes.shape[1], fill_value=np.nan)
    bin_labels[valid_mask] = np.ravel_multi_index((bin_indexes[0, valid_mask],
                                                   bin_indexes[1, valid_mask]),
                                                   dims=(n_xbins, n_ybins),
                                                   order='F')
```

In case some events are outside the desired range, their label would be a `NaN` instead of a numerical index, so when grouping the events bin by bin, none would be assigned to those events...

Reply:
ok